validator/proposer/service.go

```diff
@@ -68,10 +88,73 @@ func (p *Proposer) Stop() error {
 	return nil
 }
 
+// SaveBlockToMemory will save the block to memory so that the received
+// attestations can be aggregated in it.
+func (p *Proposer) SaveBlockToMemory(block *pbp2p.BeaconBlock) {
+	p.blockMapping[block.GetSlotNumber()] = block
+}
+
+// GetBlockFromMemory will retrieve the block to memory.
+func (p *Proposer) GetBlockFromMemory(slotnumber uint64) (*pbp2p.BeaconBlock, error) {
+	block := p.blockMapping[slotnumber]
+
+	if block == nil {
+		return nil, errors.New("block does not exist")
+	}
+
+	if block.GetSlotNumber() != slotnumber {
+		return nil, errors.New("invalid block saved in memory")
+	}
+	return block, nil
+}
+
+// DoesAttestationExist checks if an attester has already attested to a block.
+func (p *Proposer) DoesAttestationExist(attestation *pbp2p.AttestationRecord, records []*pbp2p.AttestationRecord) bool {
+	exists := false
+	for _, record := range records {
+		if bytes.Equal(record.GetAttesterBitfield(), attestation.GetAttesterBitfield()) {
+			exists = true
+			break
+		}
+	}
+	return exists
+}
+
+// AggregateAllSignatures aggregates all the signatures of the attesters. This is currently a
+// stub for now till BLS/other singature schemes are implemented.
+func (p *Proposer) AggregateAllSignatures(attestations []*pbp2p.AttestationRecord) []uint32 {
+	var agSig []uint32
+	for _, attestation := range attestations {
+		castedslice := make([]uint32, len(attestation.GetAggregateSig()))
+		for i, num := range attestation.GetAggregateSig() {
+			castedslice[i] = uint32(num)
+
+		}
+		agSig = append(agSig, castedslice...)
+	}
+	return agSig
+}
+
+// GenerateBitmask creates the attestation bitmask from all the attester bitfields in the
+// attestation records.
+func (p *Proposer) GenerateBitmask(attestations []*pbp2p.AttestationRecord) []byte {
+	bitmask := make([]byte, len(attestations[0].GetAttesterBitfield()))
+	for _, attestation := range attestations {
+		for i, chunk := range attestation.GetAttesterBitfield() {
+			bitmask[i] = bitmask[i] | chunk
+		}
+	}
+	return bitmask
+}
+
 // run the main event loop that listens for a proposer assignment.
-func (p *Proposer) run(done <-chan struct{}, client pb.ProposerServiceClient) {
+func (p *Proposer) run(delayChan <-chan time.Time, done <-chan struct{}, client pb.ProposerServiceClient) {
```

Comment:
Likewise feedback with this run method. The amount of responsibility here is increasing signficantly and you have multiple (potentially blocking) logic routines in the single for loop.

Please break this up into smaller pieces, similiar to the feedback I recommended for Attester